rvc-implementation_test.go

```
@@ -0,0 +1,278 @@
+	setLoggerFunc       func(logger logrus.FieldLogger)
+	createPodFunc       func(podConfig vc.PodConfig) (*vc.Pod, error)
+	deletePodFunc       func(podID string) (*vc.Pod, error)
+	startPodFunc        func(podID string) (*vc.Pod, error)
+	stopPodFunc         func(podID string) (*vc.Pod, error)
+	runPodFunc          func(podConfig vc.PodConfig) (*vc.Pod, error)
+	listPodFunc         func() ([]vc.PodStatus, error)
+	statusPodFunc       func(podID string) (vc.PodStatus, error)
+	pausePodFunc        func(podID string) (*vc.Pod, error)
+	resumePodFunc       func(podID string) (*vc.Pod, error)
+	createContainerFunc func(podID string, containerConfig vc.ContainerConfig) (*vc.Pod, *vc.Container, error)
+	deleteContainerFunc func(podID, containerID string) (*vc.Container, error)
+	startContainerFunc  func(podID, containerID string) (*vc.Container, error)
+	stopContainerFunc   func(podID, containerID string) (*vc.Container, error)
+	enterContainerFunc  func(podID, containerID string, cmd vc.Cmd) (*vc.Pod, *vc.Container, *vc.Process, error)
+	statusContainerFunc func(podID, containerID string) (vc.ContainerStatus, error)
+	killContainerFunc   func(podID, containerID string, signal syscall.Signal, all bool) error
+}
+
+// testingImpl is a concrete mock RVC implementation used for testing
+var testingImpl = &mockImpl{}
+
+// mockErrorPrefix is a string that all errors returned by the mock
+// implementation itself will contain as a prefix.
+const mockErrorPrefix = "mockImpl forced failure"
+
+func init() {
+	fmt.Printf("INFO: switching to fake virtcontainers implementation for testing\n")
+	vci = testingImpl
+}
+
+func (impl *mockImpl) SetLogger(logger logrus.FieldLogger) {
+	if impl.setLoggerFunc != nil {
+		impl.setLoggerFunc(logger)
+		return
+	}
+
+	vc.SetLogger(logger)
+}
+
+func (impl *mockImpl) CreatePod(podConfig vc.PodConfig) (*vc.Pod, error) {
+	if impl.createPodFunc != nil {
+		return impl.createPodFunc(podConfig)
+	}
+
+	if impl.forceFailure {
```

Just curious, I'm presuming the decision to call the underlying impl if it was registered before checking if we should return error was deliberate. Question then, might there be times when we want to return an error without calling the underlying function (in which case did you intend for folks to set impl.xxx to nil before they do the call, maybe?)